### PR TITLE
[LIVY-735][RSC] Fix rpc channel closed when multi clients connect to one driver

### PR DESCRIPTION
**Pull request by runzhiwang**
## What changes were proposed in this pull request?

Currently, the driver tries to support communicating with multi-clients, by registering each client at https://github.com/apache/incubator-livy/blob/master/rsc/src/main/java/org/apache/livy/rsc/driver/RSCDriver.java#L220.

But actually, if multi-clients connect to one driver, the rpc channel will close, the reason are as follows.

1.  In every communication, client sends two packages to driver: header{type, id}, and payload at https://github.com/apache/incubator-livy/blob/master/rsc/src/main/java/org/apache/livy/rsc/rpc/RpcDispatcher.java#L144. 

2. If client1 sends header1, payload1, and client2 sends header2, payload2 at the same time. 
  The driver receives the package in the order: header1, header2, payload1, payload2.

3. When driver receives header1, driver assigns lastHeader at https://github.com/apache/incubator-livy/blob/master/rsc/src/main/java/org/apache/livy/rsc/rpc/RpcDispatcher.java#L73.

4. Then driver receives header2, driver process it as a payload at https://github.com/apache/incubator-livy/blob/master/rsc/src/main/java/org/apache/livy/rsc/rpc/RpcDispatcher.java#L78 which cause exception and rpc channel closed.

In the muti-active HA mode, the design doc is at: https://docs.google.com/document/d/1bD3qYZpw14_NuCcSGUOfqQ0pqvSbCQsOLFuZp26Ohjc/edit?usp=sharing, the session is allocated among servers by consistent hashing. If a new livy joins, some session will be migrated from old livy to new livy. If the session client in new livy connect to driver before stoping session client in old livy, then two session clients will both connect to driver, and rpc channel close.  In this case, it's hard to ensure only one client connect to one driver at any time. So it's better to support multi-clients connect to one driver, which has no side effects.

How to fix:
1. Move the code of processing client message from `RpcDispatcher` to each `Rpc`.
2. Each `Rpc` registers itself to `channelRpc` in RpcDispatcher.
3. `RpcDispatcher` dispatches each message to `Rpc` according to  `ctx.channel()`.

## How was this patch tested?

Existed UT and IT

### CODE REVIEW

**Comment by runzhiwang** [comment]
@jerryshao @vanzin @harishreedharan @yiheng Clould you help to review ?

**Comment by tprelle** [comment]
@runzhiwang working well thanks. Your patch solve our issue between spark-magic and livy.
During our stress test we found another one I think link to this bug.

Kerberos handshake are in 2 phase : client send a first request, server respond 401 asking for the kerberos ticket then client send his kerberos ticket, server respond 200 or 201.

Then issue is : 
 1) client 1 send a  request, server respond 401.
 2) During the response, client 2 send also a  request, server respond 401.
 3) Client 1 send his kerberos ticket, it get denied by the server waiting for the client 2 kerberos ticket not the client 1
 4) Client 2 send his kerberos ticket, server respond 200

The log from request log

```
ip - - [23/Dec/2019:15:06:49 +0000] "GET https://url:8999/sessions/628/statements/8 HTTP/1.1" 401 356
ip - - [23/Dec/2019:15:06:49 +0000] "GET https://url:8999/sessions/628/statements/8 HTTP/1.1" 200 177
ip - - [23/Dec/2019:15:06:51 +0000] "POST https://url:8999/sessions/630/statements HTTP/1.1" 401 354
ip - - [23/Dec/2019:15:06:51 +0000] "POST https://url:8999/sessions/629/statements HTTP/1.1" 401 354
ip - - [23/Dec/2019:15:06:51 +0000] "POST https://url:8999/sessions/629/statements HTTP/1.1" 201 102
ip - - [23/Dec/2019:15:06:52 +0000] "GET https://url:8999/sessions/629/statements/4 HTTP/1.1" 401 356
ip - - [23/Dec/2019:15:06:52 +0000] "GET https://url:8999/sessions/629/statements/4 HTTP/1.1" 200 177
ip - - [23/Dec/2019:15:06:54 +0000] "POST https://url:8999/sessions/628/statements HTTP/1.1" 401 354
ip - - [23/Dec/2019:15:06:54 +0000] "POST https://url:8999/sessions/628/statements HTTP/1.1" 201 101
ip - - [23/Dec/2019:15:06:55 +0000] "GET https://url:8999/sessions/628/statements/9 HTTP/1.1" 401 356
ip - - [23/Dec/2019:15:06:55 +0000] "GET https://url:8999/sessions/628/statements/9 HTTP/1.1" 200 177
ip - - [23/Dec/2019:15:06:56 +0000] "POST https://url:8999/sessions/630/statements HTTP/1.1" 401 354
ip - - [23/Dec/2019:15:06:56 +0000] "POST https://url:8999/sessions/630/statements HTTP/1.1" 201 99
```

Do you think, it's link ?

**Comment by runzhiwang** [comment]
@tprelle Do your client1 and client2 connect to the same driver ?

**Comment by tprelle** [comment]
@runzhiwang same livy but not the same session so not the same driver. 

ok thanks

**Comment by runzhiwang** [comment]
@tprelle Hi, Could you explain why this patch solve your issue between spark-magic and livy?

**Comment by tprelle** [comment]
@runzhiwang We are using spark-magic inside jupyterhub, using just one livy
Without your patch,  if two users on the same host use two differents notebook in parallel it's working well. But when the same user  on the same host use more than one notebook (so same user but one unique session by notebook) in parallel. One notebook is working well other get 'Error 500 RPC channel will close'. 

With your patch same user can use notebook in parallel. 
Maybe your patch fix more than you think ?

**Comment by runzhiwang** [comment]
@tprelle Okay, I see. Thank you very much.

File: rsc/src/main/java/org/apache/livy/rsc/rpc/Rpc.java [hunk 1]
```diff
@@ -238,6 +244,166 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
     });
   }
 
+  /**
+   * Override this to add a name to the dispatcher, for debugging purposes.
```
Comment (jerryshao) [inline rsc/src/main/java/org/apache/livy/rsc/rpc/Rpc.java]: IIUC, I don't think this method overrides anything, am I right?

File: rsc/src/main/java/org/apache/livy/rsc/rpc/Rpc.java [hunk 2]
```diff
@@ -238,6 +244,166 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
...
+   * @return The name of this dispatcher.
+   */
+  protected String name() {
+    return getClass().getSimpleName();
+  }
+
+  public void handleMsg(ChannelHandlerContext ctx, Object msg, Class<?> handleClass, Object obj)
+          throws Exception {
```
Comment (jerryshao) [inline rsc/src/main/java/org/apache/livy/rsc/rpc/Rpc.java]: 4 space indent.

File: rsc/src/main/java/org/apache/livy/rsc/rpc/Rpc.java [hunk 3]
```diff
@@ -238,6 +244,166 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
...
+    if (lastHeader == null) {
+      if (!(msg instanceof MessageHeader)) {
+        LOG.warn("[{}] Expected RPC header, got {} instead.", name(),
+                msg != null ? msg.getClass().getName() : null);
```
Comment (jerryshao) [inline rsc/src/main/java/org/apache/livy/rsc/rpc/Rpc.java]: 2 space indent.

File: rsc/src/main/java/org/apache/livy/rsc/rpc/Rpc.java [hunk 4]
```diff
@@ -238,6 +244,166 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
...
+        throw new IllegalArgumentException();
+      }
+      lastHeader = (MessageHeader) msg;
+    } else {
+      LOG.debug("[{}] Received RPC message: type={} id={} payload={}", name(),
+              lastHeader.type, lastHeader.id, msg != null ? msg.getClass().getName() : null);
```
Comment (jerryshao) [inline rsc/src/main/java/org/apache/livy/rsc/rpc/Rpc.java]: ditto.

File: rsc/src/main/java/org/apache/livy/rsc/rpc/Rpc.java (hunk 1 above)
Comment (runzhiwang) [reply rsc/src/main/java/org/apache/livy/rsc/rpc/Rpc.java]: Yes, you are right. This Override  comment was copied from RpcDispatcher.java, which also does not override anything. I will remove it.

**Comment by runzhiwang** [comment]
@jerryshao Has fixed the the indention issue.

**Comment by jerryshao** [comment]
LGTM, merging to master branch.

**Pull request closed without merging**